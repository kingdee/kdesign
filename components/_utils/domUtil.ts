import React from 'react'
import scrollIntoView from 'dom-scroll-into-view'

export const scrollToDirectory = (ref: React.MutableRefObject<any>) => {
  const directoryElement: HTMLDivElement = ref.current
  if (directoryElement) {
    const rect = ref.current.getBoundingClientRect()
    const parent = ref.current.parentElement
    const parentRect = parent.getBoundingClientRect()

    const relativeTop = rect.top - parentRect.top
    const relativeBottom = rect.bottom - parentRect.top
    const isHide = relativeTop < 0 || relativeBottom > parent.clientHeight

    if (isHide) {
      scrollIntoView(ref.current, ref.current.parentElement)
    }
  }
}
