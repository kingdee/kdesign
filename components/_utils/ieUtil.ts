export const isIE = (window as any)?.ActiveXObject || 'ActiveXObject' in window

export const getFileUtil = (file: any, name = '') => {
  if (file) {
    if (isIE) {
      const blob: any = new Blob([file], { type: file.type })
      blob.fileName = name || file.name
      blob.lastModifiedDate = file?.lastModifiedDate || Date.now()
      return blob
    } else {
      return new File([file], name || file.name, { type: file.type })
    }
  }

  return file
}
