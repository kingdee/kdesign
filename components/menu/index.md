---
category: Components
type: 导航
order: 2
title: Menu
subtitle: 导航菜单
---

导航菜单是一种能快速导航到目标页面的组件。

## 使用场景

需要引导用户查找功能，避免用户迷失时。

## API

### Menu

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| additionalTools | 底部附加工具  | ReactNode  | - | - | 1.0.0 |
| theme | 菜单主题颜色 | string | `dark` | `light`  `dark` | 1.0.0 |
| mode | 菜单类型，支持垂直和内嵌模式 | string | `vertical` | `vertical` `inline` | 1.0.0 |
| collapsed | 菜单是否收起状态 | boolean | `false` | `true` `false` | 1.0.0 |
| accordion | 手风琴模式，仅在`mode='inline'`下有效  | boolean | `false` | `true` `false` | 1.5.0 |
| defaultOpenKeys | 初始展开的 SubMenu 菜单项 key 数组  | string[] | - | - | 1.0.0 |
| defaultSelectedKey | 初始选中的菜单项 | string | - | - | 1.0.0 |
| forceSubMenuRender | 在子菜单展示之前就渲染进 DOM | boolean | `false` | `true` `false` | 1.0.0 |
| inlineIndent | 菜单收起时的宽度 | number | `50` | - | 1.0.0 |
| openKeys | 当前展开的 SubMenu 菜单项 key 数组  | string[] | - | - | 1.0.0 |
| selectedKey | 当前选中的菜单项  | string | - | - | 1.0.0 |
| triggerSubMenuAction | SubMenu 展开/关闭的触发行为。内嵌模式固定为click，即该值设置只对垂直模式有效 | string | `hover` | `hover` `click` | 1.0.0 |
| onClick | 点击 MenuItem 时的回调  | function({ item, key, keyPath, domEvent })  | - | - | 1.0.0 |
| onOpenChange | SubMenu 展开/关闭的回调  | function(openKey: string)   | - | - | 1.0.0 |

### Menu.SubMenu

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| disabled | 是否禁用 | boolean | `false` | `true` `false` | 1.0.0 |
| icon | 菜单图标 | ReactNode | - | - | 1.0.0 |
| key | item的唯一标识 | string \| number | - | - | 1.0.0 |
| popupClassName | 子菜单样式 | string | - | - | 1.0.0 |
| popupOffset | 子菜单偏移量(inline模式下无效) | \[number, number\] | 数组第一个元素值为left，第二个为top | - | 1.0.0 |
| title | 子菜单标题 | string | - | - | 1.0.0 |

### Menu.Item

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| disabled | 是否禁用 | boolean | `false` | `true` `false` | 1.0.0 |
| icon | 菜单图标 | ReactNode | - | - | 1.0.0 |
| key | item的唯一标识 | string \| number | - | - | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-menu-color-disabled | --kd-g-color-disabled | #b2b2b2 |
|  | --kd-c-menu-sub-color-text | - | rgba(255, 255, 255, 0.65) |
|  | --kd-c-menu-sub-color-text-hover | --kd-g-color-white | #fff |
|  | --kd-c-menu-sub-color-text-active | --kd-g-color-white | #fff |
|  | --kd-c-menu-sub-color-background | - | #121319 |
|  | --kd-c-menu-sub-inline-color-background | - | #1f212b |
|  | --kd-c-menu-inline-color-active | --kd-g-color-white | #fff |
|  | --kd-c-menu-color-background | - | #343848 |
|  | --kd-c-menu-light-color-text | --kd-g-color-text-primary | #212121 |
|  | --kd-c-menu-light-color-text-hover | --kd-g-color-text-primary | #212121 |
|  | --kd-c-menu-light-color-text-active | --kd-g-color-theme | #5582f3 |
|  | --kd-c-menu-light-color-background-hover | - | #f5f5f5 |
|  | --kd-c-menu-light-color-background-active | - | #e3ebff |
| font | --kd-c-menu-icon-font-size | --kd-g-font-size-xx-large | 20px |
| motion | --kd-c-menu-motion-duration | --kd-g-duration | 0.3s |
| sizing | --kd-c-menu-item-sizing-height | - | 50px |
|  | --kd-c-menu-sizing-max-width | - | - |
|  | --kd-c-menu-sizing-min-width | - | - |
| z-index | --kd-c-menu-z-index | --kd-g-z-index-dialog | 1050 |
