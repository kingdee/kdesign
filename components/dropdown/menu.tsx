import * as React from 'react'
import { ConfigContext } from '../config-provider'
import { getCompProps } from '../_utils'
import classNames from 'classnames'

export interface MenuProps {
  selectedKey?: string
  className?: string
  selectable?: boolean
  style?: React.CSSProperties
  children?: React.ReactNode
  onClick?: (key: string) => void
}

export interface ItemProps {
  key?: string | number
  danger?: boolean
  divided?: boolean
  disabled?: boolean
  selected?: boolean
  className?: string
  children?: React.ReactNode
  style?: React.CSSProperties
}

const Menu = React.forwardRef<unknown, MenuProps>((props, ref) => {
  const { getPrefixCls, prefixCls: pkgPrefixCls, compDefaultProps: userDefaultProps } = React.useContext(ConfigContext)

  // 属性需要合并一遍用户定义的默认属性
  const {
    prefixCls: customPrefixcls,
    children,
    selectable,
    selectedKey,
    className,
    ...restProps
  } = getCompProps('DropdownMenu', userDefaultProps, props)

  // className前缀
  const prefixCls = getPrefixCls!(pkgPrefixCls, 'dropdown-menu', customPrefixcls)

  const cloneItem = (item: React.ReactElement, index?: number) => {
    const key = item.key || index
    const selected = selectable && String(selectedKey) === String(key)
    return React.cloneElement(item, { key, selected, defaultKey: key })
  }

  return (
    <ul className={classNames(prefixCls, className)} role="menu" {...restProps} ref={ref}>
      {React.Children.map(children, cloneItem)}
    </ul>
  )
})

const Item: React.FC<ItemProps> = (props) => {
  const { getPrefixCls, prefixCls: pkgPrefixCls, compDefaultProps: userDefaultProps } = React.useContext(ConfigContext)

  // 属性需要合并一遍用户定义的默认属性
  const {
    prefixCls: customPrefixcls,
    danger,
    divided,
    disabled,
    children,
    selected,
    defaultKey,
    className,
    ...restProps
  } = getCompProps('DropdownItem', userDefaultProps, props)

  // className前缀
  const prefixCls = getPrefixCls!(pkgPrefixCls, 'dropdown-menu-item', customPrefixcls)

  const childrenIsLink = children.type === 'a'

  const linkChildren = <span>{children.props?.children}</span>

  const itemRef = React.useRef()

  return (
    <li
      ref={itemRef}
      data-key={defaultKey}
      title={(itemRef.current as any)?.textContent}
      className={classNames(prefixCls, className, {
        danger,
        divided,
        disabled,
        selected,
      })}
      role="menuitem"
      {...restProps}
    >
      {childrenIsLink ? (
        disabled ? (
          linkChildren
        ) : (
          React.cloneElement(children, { children: linkChildren })
        )
      ) : (
        <span>{children}</span>
      )}
    </li>
  )
}

Menu.displayName = 'DropdownMenu'
Item.displayName = 'DropdownMenuItem'

export { Menu, Item }
