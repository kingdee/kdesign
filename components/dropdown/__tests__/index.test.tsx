import React from 'react'
import { mount } from 'enzyme'
import Dropdown from '../index'
import { Placements, Triggers } from '../../_utils/usePopper'
import { Menu, Item } from '../menu'
import mountTest from '../../../tests/shared/mountTest'

const menu = (
  <Menu>
    <Item>
      <a target="_blank" rel="noopener noreferrer" href="https://www.kingdee.com/">
        菜单1
      </a>
    </Item>
    <Item>
      <a target="_blank" rel="noopener noreferrer" href="https://www.kdcloud.com/">
        菜单2
      </a>
    </Item>
    <Item>
      <a target="_blank" rel="noopener noreferrer" href="https://www.yunzhijia.com/">
        菜单3
      </a>
    </Item>
    <Item danger>a danger item</Item>
  </Menu>
)

const wrapperRef = React.createRef() as any

describe('Dropdown', () => {
  Placements.forEach((placement) => {
    mountTest(() => (
      <Dropdown menu={menu} placement={placement}>
        <span />
      </Dropdown>
    ))
  })

  it('renders correctly', () => {
    const wrapper = mount(
      <Dropdown menu={menu}>
        <span />
      </Dropdown>,
    )
    expect(wrapper).toMatchSnapshot()

    Placements.forEach((placement) => {
      expect(
        mount(
          <Dropdown menu={menu} placement={placement}>
            <span />
          </Dropdown>,
        ),
      ).toMatchSnapshot()
    })

    Triggers.forEach((trigger) => {
      expect(
        mount(
          <Dropdown menu={menu} trigger={[trigger]}>
            <span />
          </Dropdown>,
        ),
      ).toMatchSnapshot()
    })
  })

  it('menu is Array', () => {
    const wrapper = mount(
      <Dropdown menu={[]}>
        <button type="button">button</button>
      </Dropdown>,
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should have displayName static property', () => {
    const wrapper = mount(
      <Dropdown menu={menu}>
        <span />
      </Dropdown>,
    )
    expect((wrapper.type() as any).displayName).toBe('Dropdown')
  })

  it('should not visible when dropdown is disabled', () => {
    const onVisibleChange = jest.fn()
    const wrapper = mount(
      <Dropdown menu={menu} onVisibleChange={onVisibleChange} disabled>
        <span />
      </Dropdown>,
    )
    wrapper.simulate('mouseenter')
    expect(onVisibleChange).not.toHaveBeenCalledWith()
  })

  it('should visible', () => {
    const onClick = jest.fn()
    const wrapper = mount(
      <Dropdown menu={menu} trigger="click">
        <span onClick={onClick} />
      </Dropdown>,
    )
    wrapper.find('.kd-dropdown-trigger').simulate('click')
    expect(onClick).toHaveBeenCalled()
  })

  it('should menu item onItemClick', () => {
    const onItemClick = jest.fn()

    const wrapper = mount(
      <div ref={wrapperRef}>
        <Dropdown
          visible
          menu={menu}
          trigger={'click'}
          onItemClick={onItemClick}
          getPopupContainer={() => wrapperRef.current}
        >
          <span />
        </Dropdown>
      </div>,
    )
    wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(0).simulate('click')
    expect(onItemClick).toHaveBeenCalled()
  })

  it('should menu has arrow', () => {
    const wrapper = mount(
      <div ref={wrapperRef}>
        <Dropdown arrow visible menu={menu} trigger={'click'} getPopupContainer={() => wrapperRef.current}>
          <span />
        </Dropdown>
      </div>,
    )
    expect(wrapper.find('.kd-dropdown')).toHaveClassName('arrow')
    expect(wrapper.find('.kd-dropdown')).toHaveStyle('--arrowSize', '4.25px')
  })

  it('should menu item disabled', () => {
    const onItemClick = jest.fn()
    const menu = (
      <Menu>
        <Item disabled>
          <a target="_blank" rel="noopener noreferrer" href="https://www.kingdee.com/">
            菜单1
          </a>
        </Item>
      </Menu>
    )

    const wrapper = mount(
      <div ref={wrapperRef}>
        <Dropdown
          visible
          menu={menu}
          trigger={'click'}
          onItemClick={onItemClick}
          getPopupContainer={() => wrapperRef.current}
        >
          <span />
        </Dropdown>
      </div>,
    )
    wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(0).simulate('click')
    expect(onItemClick).not.toHaveBeenCalledWith('菜单1')
  })

  it('should menu item onClick', () => {
    const onClick = jest.fn()
    const menu = (
      <Menu onClick={onClick}>
        <Item key="one">
          <a target="_blank" rel="noopener noreferrer" href="https://www.kdcloud.com/">
            菜单1
          </a>
        </Item>
      </Menu>
    )

    const wrapper = mount(
      <div ref={wrapperRef}>
        <Dropdown visible menu={menu} trigger={'click'} getPopupContainer={() => wrapperRef.current}>
          <span />
        </Dropdown>
      </div>,
    )
    wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(0).simulate('click')
    expect(onClick).toHaveBeenCalledWith('one')
  })

  it('should menu item danger and divided', () => {
    const menu = (
      <Menu>
        <Item danger>
          <a target="_blank" rel="noopener noreferrer" href="https://www.kingdee.com/">
            菜单1
          </a>
        </Item>
        <Item divided>
          <a target="_blank" rel="noopener noreferrer" href="https://www.kdcloud.com/">
            菜单2
          </a>
        </Item>
      </Menu>
    )

    const wrapper = mount(
      <div ref={wrapperRef}>
        <Dropdown visible menu={menu} trigger={'click'} getPopupContainer={() => wrapperRef.current}>
          <span />
        </Dropdown>
      </div>,
    )
    expect(wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(0)).toHaveClassName('danger')
    expect(wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(1)).toHaveClassName('divided')
  })

  it('should menu item can select able', () => {
    const wrapper = mount(
      <div ref={wrapperRef}>
        <Dropdown arrow visible menu={menu} trigger={'click'} getPopupContainer={() => wrapperRef.current} selectable>
          <span />
        </Dropdown>
      </div>,
    )
    wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(0).simulate('click')
    expect(wrapper.find('.kd-dropdown-menu').find('.kd-dropdown-menu-item').at(0)).toHaveClassName('selected')
  })
})
