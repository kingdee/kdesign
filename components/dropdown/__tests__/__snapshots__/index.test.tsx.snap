// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown menu is Array 1`] = `
<Dropdown
  menu={Array []}
>
  <button
    className="kd-dropdown-trigger"
    type="button"
  >
    button
  </button>
</Dropdown>
`;

exports[`Dropdown renders correctly 1`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 2`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="top"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 3`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="left"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 4`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="right"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 5`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="bottom"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 6`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="topLeft"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 7`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="topRight"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 8`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="bottomLeft"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 9`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="bottomRight"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 10`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="leftTop"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 11`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="leftBottom"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 12`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="rightTop"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 13`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  placement="rightBottom"
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 14`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  trigger={
    Array [
      "hover",
    ]
  }
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 15`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  trigger={
    Array [
      "focus",
    ]
  }
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 16`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  trigger={
    Array [
      "click",
    ]
  }
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;

exports[`Dropdown renders correctly 17`] = `
<Dropdown
  menu={
    <DropdownMenu>
      <DropdownMenuItem>
        <a
          href="https://www.kingdee.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单1
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.kdcloud.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单2
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem>
        <a
          href="https://www.yunzhijia.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
          菜单3
        </a>
      </DropdownMenuItem>
      <DropdownMenuItem
        danger={true}
      >
        a danger item
      </DropdownMenuItem>
    </DropdownMenu>
  }
  trigger={
    Array [
      "contextMenu",
    ]
  }
>
  <span
    className="kd-dropdown-trigger"
  />
</Dropdown>
`;
